### DISCUSSION TITLE
How To Accept User Input?

### DISCUSSION THREAD

**Discussion by TheCedarPrince** [FedeClaudi/Term.jl#133] [General]
Hi @FedeClaudi !

How do I use Term.jl to accept user input?
For example, I have the following render: 

![image](https://user-images.githubusercontent.com/29561456/177026662-0a601104-bcc7-4537-8081-99d20a22ca40.png)

and would somehow like to use `prompt` within it.
When I tried using prompt inside of a `Panel`, I get this:

![image](https://user-images.githubusercontent.com/29561456/177026685-92e2ec9b-08e8-4260-ae42-3b7c0ab2be33.png)

How do I accept user input within the box?
Any thoughts?
Thank you!

**Comment by FedeClaudi** [comment]
Hi, 

Currently you can't unfortunately, I haven't tackled inputs yet.

**Comment by johnnychen94** [comment]
I believe the missing key component here is a new Panel-like object with generic `IO` support. Our GSoC project would also require this when trying to assembled REPL with Term.jl https://github.com/soumitradev/TermLayouts.jl

cc: @soumitradev

**Comment by TheCedarPrince** [comment]
Ah thanks @FedeClaudi and @johnnychen94 .
I was wondering if this was possible.
Seems a tricky problem as I could see two ways of approaching it:

1. Give a new Panel object that is wrapped around `prompt` a fixed amount of space for typing into.
2. Redraw the box (i.e. buffer) around someone's input each time they input a character into this new PanelInput object or something like that.

I could see it being quite challenging! 
I see that rich does support a form of it here: https://rich.readthedocs.io/en/latest/_modules/rich/console.html#Console.input

**Comment by FedeClaudi** [comment]
Hey, I've converted this to a discussion since we're throwing ideas around and I won't get to implementing them any time soon.

I see two challenges:
1. I'd need to look into how `prompt` works, but likely we'd have to use ANSI commands to place the cursor location "inside" a panel instead of at the end of the last line
2. The panel should re-size automatically as the size of the input changes.

For (2), I haven't implemented this kind of "live" updatable renderables yet. It likely would just be a panel wrapped into something that refreshes the display at, say, 60 Hz. A bit expensive to create a new panel at each refresh but it might be ok performance wise.


----

Rich does have `Prompt` [classes](https://rich.readthedocs.io/en/latest/prompt.html) but they don't wrap it around a Panel. They just add some style and some pre-defined prompt texts. That would be much easier to do.

**Comment by FedeClaudi** [comment]
Just an update to say that `v2.0` will include a `Prompt` renderable for getting user input: https://github.com/FedeClaudi/Term.jl/pull/175

**Reply by TheCedarPrince** [reply]
Amazing! I was thinking about Term.jl recently and am excited to see about bringing it into my tools! :smile:
